src/logics/magicsea/logic.swap-token.ts

```diff
@@ -0,0 +1,294 @@
+
+      return {
+        tradeType,
+        input,
+        output,
+        path,
+        priceImpact,
+        slippage,
+      };
+    }
+  }
+
+  async build(fields: SwapTokenLogicFields, options: SwapTokenLogicOptions) {
+    const { tradeType, input, output, path, slippage, balanceBps } = fields;
+    const { account } = options;
+
+    const receiver = await this.calcAgent(account);
+    const deadline = BigNumber.from(Math.floor(Date.now() / 1000)).add(1800); // 30m
+
+    let data, inputs;
+    if (tradeType === core.TradeType.exactIn) {
+      const amountIn = input.amountWei;
+      const amountOutMin = slippage ? common.calcSlippage(output.amountWei, slippage) : output.amountWei;
+      data = MagicSeaRouter__factory.createInterface().encodeFunctionData('swapExactTokensForTokens', [
+        amountIn,
+        amountOutMin,
+        path,
+        receiver,
+        deadline,
+      ]);
+
+      const amountOffset = balanceBps ? common.getParamOffset(0) : undefined;
+      inputs = [
+        core.newLogicInput({
+          input: new common.TokenAmount(input.token.wrapped, input.amount),
+          balanceBps,
+          amountOffset,
+        }),
+      ];
+
+      const wrapMode = input.token.isNative
+        ? core.WrapMode.wrapBefore
+        : output.token.isNative
+        ? core.WrapMode.unwrapAfter
+        : core.WrapMode.none;
+
+      return core.newLogic({ to: getContractAddress(this.chainId, 'Router'), data, inputs, wrapMode });
+    } else {
+      const amountInMax = input.amountWei;
+      const amountOut = output.amountWei;
+      data = MagicSeaRouter__factory.createInterface().encodeFunctionData('swapTokensForExactTokens', [
+        amountOut,
+        amountInMax,
+        path,
+        receiver,
+        deadline,
+      ]);
+
+      inputs = [
+        core.newLogicInput({
+          input: new common.TokenAmount(input.token.wrapped, input.amount),
+        }),
+      ];
+
+      const wrapMode = input.token.isNative
+        ? core.WrapMode.wrapBefore
+        : output.token.isNative
+        ? core.WrapMode.unwrapAfter
+        : core.WrapMode.none;
+
+      return core.newLogic({ to: getContractAddress(this.chainId, 'Router'), data, inputs, wrapMode });
+    }
+  }
+
+  async getAllCommonPairs(tokenA: Token, tokenB: Token, isAll = true): Promise<Pair[]> {
+    const pairTokens = this.generateAllRoutePairs(tokenA, tokenB);
+    const pairAddresses = await this.getPairAddresses(pairTokens);
+    const addresses = isAll ? pairAddresses : [pairAddresses[0]];
+
+    const results: [BigintIsh, BigintIsh][] = [];
+    for (let i = 0; i < addresses.length; i++) {
+      const pairAddress = addresses[i];
+      if (!pairAddress) {
+        results.push(['0', '0']);
+      } else {
+        const pair = MagicSeaPair__factory.connect(pairAddress, this.provider);
+        const [reserve0, reserve1] = await pair.getReserves();
+        results.push([reserve0.toString(), reserve1.toString()]);
+      }
+    }
```

Comment:
should we consider using Multicall3 to reduce the number of contract calls. Currently, one `quote` will trigger 16 contract calls? 
8 times for `getPairAddresses`
8 times for `await pair.getReserves();`

Reply:
IMO, using multi-call to reduce calls is great but the logic would become more complicated. Whether to do the optimization or not depends on the RPC capability of the IOTA chain. What do @zodahu think?

Reply:
I think we can use multicall here as get pairs is commonly used in this way, and our uni-v3 has implemented similar code: https://github.com/dinngo/protocolink-logics/blob/master/src/modules/univ3/logic.swap-token.ts#L240-L283